### PR TITLE
Add scripts for GitHub Pages Deployment in Tutorial 4

### PR DESCRIPTION
**Pull request by ymekuria**
solution checks if the project is used locally only or is deployed in production
script CSS asset URLs require the repo name to be prepended to deploy to GitHub Pages correctly

package.json 

`npm run deploy` runs the script along with other commands to deploy to GitHub Pages

### CODE REVIEW

File: examples/zkapps/04-zkapp-browser-ui/ui/ghp-postbuild.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  files.forEach(file => {
+    if (path.extname(file) === '.css') {
+      const filePath = path.join(cssDir, file);
+      fs.readFile(filePath, 'utf8', (err, data) => {
+        if (err) {
+          console.error(err);
+          process.exit(1);
+        }
+        
+        // 04-zkapp-browser-ui should be replaced with your repo name if it is different.
```
Comment (barriebyron) [inline examples/zkapps/04-zkapp-browser-ui/ui/ghp-postbuild.js:23]:
```suggestion
        // replace 04-zkapp-browser-ui with your repo name if it is different.
```

File: examples/zkapps/04-zkapp-browser-ui/ui/out/404.html [hunk 2]
```diff
@@ -0,0 +1 @@
.../IbSPwN67V1Yvix1r0yGn1/_ssgManifest.js" defer=""></script></head><body><div id="__next"><div style="font-family:system-ui,&quot;Segoe UI&quot;,Roboto,Helvetica,Arial,sans-serif,&quot;Apple Color Emoji&quot;,&quot;Segoe UI Emoji&quot;;height:100vh;text-align:center;display:flex;flex-direction:column;align-items:center;justify-content:center"><div style="line-height:48px"><style>body{color:#000;background:#fff;margin:0}.next-error-h1{border-right:1px solid rgba(0,0,0,.3)}@media (prefers-color-scheme:dark){body{color:#fff;background:#000}.next-error-h1{border-right:1px solid rgba(255,255,255,.3)}}</style><h1 class="next-error-h1" style="display:inline-block;margin:0 20px 0 0;padding-right:23px;font-size:24px;font-weight:500;vertical-align:top">404</h1><div style="display:inline-block"><h2 style="font-size:14px;font-weight:400;line-height:28px">This page could not be found<!-- -->.</h2></div></div></div></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"statusCode":404}},"page":"/_error","query":{},"buildId":"IbSPwN67V1Yvix1r0yGn1","assetPrefix":"/04-zkapp-browser-ui","nextExport":true,"isFallback":false,"gip":true,"scriptLoader":[]}</script></body></html>
```
Comment (barriebyron) [inline examples/zkapps/04-zkapp-browser-ui/ui/out/404.html:1]:
```suggestion
<!DOCTYPE html><html><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width"/><title>404: This page could not be found. Be sure to check Tutorial 4 for required configuration.</title><meta name="next-head-count" content="3"/><link rel="preload" href="/04-zkapp-browser-ui/_next/static/css/ad2354abebf7054a.css" as="style"/><link rel="stylesheet" href="/04-zkapp-browser-ui/_next/static/css/ad2354abebf7054a.css" data-n-g=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/04-zkapp-browser-ui/_next/static/chunks/polyfills-78c92fac7aa8fdd8.js"></script><script src="/04-zkapp-browser-ui/_next/static/chunks/webpack-683523282bec2506.js" defer=""></script><script src="/04-zkapp-browser-ui/_next/static/chunks/framework-305cb810cde7afac.js" defer=""></script><script src="/04-zkapp-browser-ui/_next/static/chunks/main-15ee87a5f6d80118.js" defer=""></script><script src="/04-zkapp-browser-ui/_next/static/chunks/pages/_app-7e7a1410ed9b0062.js" defer=""></script><script src="/04-zkapp-browser-ui/_next/static/chunks/pages/_error-54de1933a164a1ff.js" defer=""></script><script src="/04-zkapp-browser-ui/_next/static/IbSPwN67V1Yvix1r0yGn1/_buildManifest.js" defer=""></script><script src="/04-zkapp-browser-ui/_next/static/IbSPwN67V1Yvix1r0yGn1/_ssgManifest.js" defer=""></script></head><body><div id="__next"><div style="font-family:system-ui,&quot;Segoe UI&quot;,Roboto,Helvetica,Arial,sans-serif,&quot;Apple Color Emoji&quot;,&quot;Segoe UI Emoji&quot;;height:100vh;text-align:center;display:flex;flex-direction:column;align-items:center;justify-content:center"><div style="line-height:48px"><style>body{color:#000;background:#fff;margin:0}.next-error-h1{border-right:1px solid rgba(0,0,0,.3)}@media (prefers-color-scheme:dark){body{color:#fff;background:#000}.next-error-h1{border-right:1px solid rgba(255,255,255,.3)}}</style><h1 class="next-error-h1" style="display:inline-block;margin:0 20px 0 0;padding-right:23px;font-size:24px;font-weight:500;vertical-align:top">404</h1><div style="display:inline-block"><h2 style="font-size:14px;font-weight:400;line-height:28px">This page could not be found<!-- -->.</h2></div></div></div></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"statusCode":404}},"page":"/_error","query":{},"buildId":"IbSPwN67V1Yvix1r0yGn1","assetPrefix":"/04-zkapp-browser-ui","nextExport":true,"isFallback":false,"gip":true,"scriptLoader":[]}</script></body></html>
```
```suggestion
<!DOCTYPE html><html><head><meta charSet="utf-8"/><meta name="viewport" content="width=device-width"/><title>404: This page could not be found</title><meta name="next-head-count" content="3"/><link rel="preload" href="/04-zkapp-browser-ui/_next/static/css/ad2354abebf7054a.css" as="style"/><link rel="stylesheet" href="/04-zkapp-browser-ui/_next/static/css/ad2354abebf7054a.css" data-n-g=""/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/04-zkapp-browser-ui/_next/static/chunks/polyfills-78c92fac7aa8fdd8.js"></script><script src="/04-zkapp-browser-ui/_next/static/chunks/webpack-683523282bec2506.js" defer=""></script><script src="/04-zkapp-browser-ui/_next/static/chunks/framework-305cb810cde7afac.js" defer=""></script><script src="/04-zkapp-browser-ui/_next/static/chunks/main-15ee87a5f6d80118.js" defer=""></script><script src="/04-zkapp-browser-ui/_next/static/chunks/pages/_app-7e7a1410ed9b0062.js" defer=""></script><script src="/04-zkapp-browser-ui/_next/static/chunks/pages/_error-54de1933a164a1ff.js" defer=""></script><script src="/04-zkapp-browser-ui/_next/static/IbSPwN67V1Yvix1r0yGn1/_buildManifest.js" defer=""></script><script src="/04-zkapp-browser-ui/_next/static/IbSPwN67V1Yvix1r0yGn1/_ssgManifest.js" defer=""></script></head><body><div id="__next"><div style="font-family:system-ui,&quot;Segoe UI&quot;,Roboto,Helvetica,Arial,sans-serif,&quot;Apple Color Emoji&quot;,&quot;Segoe UI Emoji&quot;;height:100vh;text-align:center;display:flex;flex-direction:column;align-items:center;justify-content:center"><div style="line-height:48px"><style>body{color:#000;background:#fff;margin:0}.next-error-h1{border-right:1px solid rgba(0,0,0,.3)}@media (prefers-color-scheme:dark){body{color:#fff;background:#000}.next-error-h1{border-right:1px solid rgba(255,255,255,.3)}}</style><h1 class="next-error-h1" style="display:inline-block;margin:0 20px 0 0;padding-right:23px;font-size:24px;font-weight:500;vertical-align:top">404</h1><div style="display:inline-block"><h2 style="font-size:14px;font-weight:400;line-height:28px">This page could not be found<!-- -->.</h2></div></div></div></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"statusCode":404}},"page":"/_error","query":{},"buildId":"IbSPwN67V1Yvix1r0yGn1","assetPrefix":"/04-zkapp-browser-ui","nextExport":true,"isFallback":false,"gip":true,"scriptLoader":[]}</script></body></html>
```

Comment (barriebyron) [reply examples/zkapps/04-zkapp-browser-ui/ui/out/404.html:1]: not sure why two changes show here, my feedback is to enhance our 404 error message with some helpful guidance

Comment (ymekuria) [reply examples/zkapps/04-zkapp-browser-ui/ui/out/404.html:1]: This 404 page is automatically generated by `NextJS`. I think it is a good idea to enhance the error message but we can create an issue and do it in a seperate pr.

**Review by barriebyron — APPROVED** [review]
LGTM for text, I did not test technical accuracy of code.
Just a few suggestions for wording

File: examples/zkapps/04-zkapp-browser-ui/ui/out/404.html (hunk 2 above)
Comment (barriebyron) [reply examples/zkapps/04-zkapp-browser-ui/ui/out/404.html:1]: our new issue https://github.com/o1-labs/zkapp-cli/issues/470

File: examples/zkapps/04-zkapp-browser-ui/ui/next.config.js [hunk 3]
```diff
@@ -30,16 +21,29 @@ const nextConfig = {
...
+
+  /* Used to serve the Next.js app from a subdirectory (the GitHub repo name) and
+   * assetPrefix is used to serve assets (JS, CSS, images, etc.) from that subdirectory
+   * when deployed to GitHub Pages. The assetPrefix needs to be added manually to any assets
+   * if they're not loaded by Next.js' automatic handling (for example, in CSS files or in a <img> element).
+   * The 'ghp-postbuild.js' script in this project prepends the repo name to asset urls in the built css files
+   * after runing 'npm run deploy'.
+   */
+  basePath: process.env.NODE_ENV === 'production' ? '/04-zkapp-browser-ui' : '', // update if your repo name changes for 'npm run deploy' to work successfully
+  assetPrefix: process.env.NODE_ENV === 'production' ? '/04-zkapp-browser-ui' : '' // update if your repo name changes for 'npm run deploy' to work successfully
```
[the remark this answers is not in the document]
Comment (ymekuria) [reply examples/zkapps/04-zkapp-browser-ui/ui/next.config.js:46]: This sentence may be a little misleading
"The GitHub repository name must match the zkApp project name for 'npm run deploy' to work successfully."

With this update https://github.com/o1-labs/zkapp-cli/pull/468, the cli takes the zkApp project name and adds it in the above configuration assuming that the zkApp project name is the same as the GitHub repository name.  `npm run deploy` will still work if the GitHub repository name is different from the zkApp project name, but the developer must update the assetPrefix and basePath to match the GitHub repository name.

**Pull request merged**